### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by xl402**

### CODE REVIEW

File: src/networks/build_attention.py [hunk 1]
```diff
@@ -3,86 +3,69 @@
...
-_CHR_IDX = string.ascii_lowercase
+def build_quadratic_attention_equation(rank, attn_axes):
+    batch_dims = _get_index_of_batch_dims(rank, attn_axes)
+    query_notation, value_notation = _get_query_and_value_notation(rank, batch_dims)
+    key_notation = value_notation
+    qk_product_notation = "".join([query_notation[i] for i in batch_dims] +
+                                  [query_notation[i] for i in attn_axes] +
+                                  [key_notation[i] for i in attn_axes])
+    kq_product_equation = f"{key_notation},{query_notation}->{qk_product_notation}"
+    qk_v_product_equation = f"{qk_product_notation},{value_notation}->{query_notation}"
```
Comment (xl402) [inline src/networks/build_attention.py:14]: added f string so ppl can just know what the output means without looking into the code

File: src/networks/build_attention.py [hunk 2]
```diff
@@ -3,86 +3,69 @@
 import numpy as np
 
 
-_CHR_IDX = string.ascii_lowercase
+def build_quadratic_attention_equation(rank, attn_axes):
+    batch_dims = _get_index_of_batch_dims(rank, attn_axes)
+    query_notation, value_notation = _get_query_and_value_notation(rank, batch_dims)
```
Comment (xl402) [inline src/networks/build_attention.py:8]: this code is reusable for quadratic and linear methods, so put them into one function

File: src/networks/model.py [hunk 3]
```diff
@@ -40,52 +40,48 @@ def _build_attention(self, rank):
             self._attention_axes = tuple(range(1, rank - 2))
         else:
             self._attention_axes = tuple(self._attention_axes)
-        self._dot_product_equation, self._combine_equation, attn_scores_rank = (
-             self._build_attention_equation(rank, attn_axes=self._attention_axes))
+        self._add_attention_equation(rank)
+        self._add_soft_max_and_dropout_layers()
+        if hasattr(self, '_build_normalisation_equation'):
+            self._add_normalisation_equation(rank)
```
Comment (xl402) [inline src/networks/model.py:46]: separated out these subtasks as private functions

File: src/networks/model.py [hunk 4]
```diff
@@ -40,52 +40,48 @@ def _build_attention(self, rank):
...
-        attention_scores_dropout = self._dropout_layer(
-            attention_scores, training=training)
-
+        attention_scores_dropout = self._dropout_layer(attention_scores, training=training)
         attention_output = einsum(self._combine_equation, attention_scores_dropout, value)
         return attention_output, attention_scores
 
     def linear_attention(self, query, key, value, attention_mask=None, training=None):
+        if attention_mask is not None:
+            raise(NotImplementedError('masked linear attention not implemented'))
```
Comment (xl402) [inline src/networks/model.py:72]: TODO: make linear attention support masking

File: src/networks/model.py [hunk 5]
```diff
@@ -40,52 +40,48 @@ def _build_attention(self, rank):
...
-
-        eq1, eq2, eq3 = self._normalisation_equations
-        k_ones = einsum(eq1, lifted_key, ones)
-        D = einsum(eq2, lifted_query, k_ones)
-        D = 1 / (D + 1e-6)
-        out = einsum(eq3, D, qkv)
-        return out, None
-
+        normalised_qkv = self._normalise(lifted_key, lifted_query, qkv)
+        return normalised_qkv, None
```
Comment (xl402) [inline src/networks/model.py:79]: clean up, chucked normalisation as a private function

File: test/test_performer.py [hunk 6]
```diff
@@ -36,6 +36,7 @@ def test_performer_linear_attention_approximates_quadratic_attention(inputs, num
     approx_layer = Performer(attention_method='linear', supports=1000, **kwargs)
     exact_layer = Performer(attention_method='quadratic', **kwargs)
 
+    tf.random.set_seed(42)
```
Comment (xl402) [inline test/test_performer.py:39]: test fails roughly 1 out of 100 due to randomness, seeded to make sure this doesn't happen
